### PR TITLE
JDBC implementation of ResultSet -> Table

### PR DESCRIPTION
**Pull request by nbauernfeind**
This is an implementation for odbc/jdbc ingestion under #1266. The type mapping is modeled after the enterprise solution, but uses an alternative interface to fill into chunks instead of per row setting.

### CODE REVIEW

File: extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java [hunk 1]
```diff
@@ -0,0 +1,1321 @@
...
+            final double value = doubleChunk.get(srcOffset);
+            if (value == QueryConstants.NULL_DOUBLE) {
+                stmt.setNull(parameterIndex, sqlType);
+            } else {
+                stmt.setBigDecimal(parameterIndex, new BigDecimal(value));
+            }
+        }
+    }
+
+    public static class ByteDataTypeMapping extends DataTypeMapping<Byte> {
```
Comment (rcaudy) [inline extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java:672]: Unused. Mistake?

File: extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java [hunk 2]
```diff
@@ -0,0 +1,1321 @@
...
+                ResultSet resultSet, int columnIndex, Context context) throws SQLException {
+            final WritableObjectChunk<LocalDate, Values> objectChunk = destChunk.asWritableObjectChunk();
+
+            final java.sql.Date value = resultSet.getDate(columnIndex, context.getSourceCalendar());
+            if (value == null) {
+                objectChunk.set(destOffset, null);
+            } else {
+                // simple value.toLocalDate() does not work, we need to use the specified calendar (JDBC returns DATE
+                // values as midnight in the timezone of the given calendar)
+                final Calendar sourceCalendar = (Calendar) context.getSourceCalendar().clone();
```
Comment (rcaudy) [inline extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java:745]: Clone into the context.

File: extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java [hunk 3]
```diff
@@ -0,0 +1,1321 @@
...
+                final Timestamp ts = new Timestamp(value.getMillis());
+                ts.setNanos((int) (value.getNanos() % 1_000_000_000L));
+                stmt.setTimestamp(parameterIndex, ts, context.getSourceCalendar());
+            }
+        }
+    }
+
+    public static class DateYearDataTypeMapping extends DataTypeMapping<Integer> {
+        DateYearDataTypeMapping() {
+            super(Types.DATE, Integer.class, java.sql.Date.class);
```
Comment (rcaudy) [inline extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java:832]: Should we be passing deephaven types that are boxed?

File: extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java [hunk 4]
```diff
@@ -0,0 +1,1321 @@
...
+     * specified, a default will be used.
+     *
+     * @param connClass a JDBC Connection subclass, used to determine database-specific mappings
+     * @param type the SQL type
+     * @param deephavenDataType the column type, if known. If null, the default type mapping will be provided
+     * @param throwException if true, throw an exception if a mapping cannot be found. Otherwise, return null
+     * @param <T> the target Deephaven type
+     * @return a mapping object that can be used to map SQL/JDBC ResultSet values to and from Deephaven column values
+     */
+    static public <T> DataTypeMapping<T> getColumnTypeMapping(final Class<? extends Connection> connClass, int type,
```
Comment (rcaudy) [inline extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java:1275]: You mentioned wanting to use the one that throw an exception. Return null on no exception should be specified on all of these.

File: extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcToTableAdapter.java [hunk 5]
```diff
@@ -7,6 +7,8 @@
 import io.deephaven.chunk.WritableChunk;
 import io.deephaven.chunk.attributes.Values;
 import io.deephaven.chunk.util.pools.ChunkPoolConstants;
```
Comment (rcaudy) [inline extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcToTableAdapter.java:9]: You have a couple of unused imports here.

File: extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java [hunk 6]
```diff
@@ -669,37 +638,9 @@ public void bindFromChunk(Chunk<Values> srcChunk, int srcOffset, PreparedStateme
...
-            } else {
-                stmt.setByte(parameterIndex, value);
-            }
-        }
-    }
-
     public static class BooleanDataTypeMapping extends DataTypeMapping<Boolean> {
         BooleanDataTypeMapping(int sqlType) {
-            super(sqlType, Boolean.class);
+            super(sqlType, boolean.class);
```
Comment (rcaudy) [inline extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java:643]: Not so sure about this change. We always used boxed `Boolean` in Deephaven columns.

File: extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java [hunk 7]
```diff
@@ -986,111 +927,108 @@ MappingCollection build() {
      */
     private static final Map<Integer, MappingCollection> dataTypeMappings;
     static {
-        final Map<Integer, MappingCollection> initMap = new HashMap<>(25);
-        initMap.put(Types.CHAR, MappingCollection.builder(String.class)
+
+        dataTypeMappings = Map.ofEntries(Map.entry(Types.CHAR, MappingCollection.builder(String.class)
```
Comment (rcaudy) [inline extensions/jdbc/src/main/java/io/deephaven/jdbc/JdbcTypeMapper.java:931]: If you added a newline between `Map.ofEntries(` and `Map.entry(` and then auto-format, you would achieve pleasing symmetry with the rest of the block.

**Review by rcaudy — COMMENTED** [review]
Looking good. Minor comments only.

**Comment by rcaudy** [comment]
What's has your testing been for this PR?

**Comment by nbauernfeind** [comment]
Testing strategy has been unit tests + manual imports of a sqlite database.

**Review by rcaudy — APPROVED** [review]
Looks like testing coverage is good for all non-vendor-specific ingest (`bindToChunk`). `bindFromChunk` is untested, but also unused until we add export capabilities.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**